pkg/virt-controller/services/template.go

```diff
@@ -1135,30 +695,12 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
 		requestResource(&resources, SevDevice)
 	}
 
-	computeContainerOpts := []Option{
-		WithVolumeDevices(volumeDevices...),
-		WithVolumeMounts(volumeMounts...),
-		WithResourceRequirements(resources),
-		WithPorts(vmi),
-		WithCapabilities(vmi),
-	}
-	if nonRoot {
-		computeContainerOpts = append(computeContainerOpts, WithNonRoot(userId))
-	}
-	if t.IsPPC64() {
-		computeContainerOpts = append(computeContainerOpts, WithPrivileged())
-	}
-	if vmi.Spec.ReadinessProbe != nil {
-		computeContainerOpts = append(computeContainerOpts, WithReadinessProbe(vmi))
-	}
-
-	if vmi.Spec.LivenessProbe != nil {
-		computeContainerOpts = append(computeContainerOpts, WithLivelinessProbe(vmi))
+	volumeRenderer, err := t.newVolumeRenderer(vmi, namespace, requestedHookSidecarList)
```

Comment:
I can see that we gather all volume requirements and create all volumes here. And now I see that the alternative style of  builder, with possible execution `builder.addSomething()` might be useful here. 

For example requestedHookSidecarList logic is divided into three places.
One is unmarshalling at line 605 `	requestedHookSidecarList, err := hooks.UnmarshalHookSidecarList(vmi)`, second is adding volumes here at line 698 `volumeRenderer, err := t.newVolumeRenderer(vmi, namespace, requestedHookSidecarList)`, and thirds at line 800 `	for i, requestedHookSidecar := range requestedHookSidecarList {`. Similarly `HugePages`. I understand that this is not bad. It is one way of doing things. We now have all volumes in one place. 

Other way would be to divide the code by some ability (like for example sidecars, hugepages). So if someone adds HookSidecars then all the code would be in one place:

```
	requestedHookSidecarList, err := hooks.UnmarshalHookSidecarList(vmi)
	volumeRenderer.addSidecarVolumes(requestedHookSidecarList)
	containersRenderer.addSidecarContainers(requestedHookSidecarList)
```
and it is also possible to add a higher level abstraction vmTemplateRenderer.withSidecars().

But, I digress. The decision was already made. ;)

Reply:
This is good feedback - my goal is to continue exporting this discombobulated code into the appropriate patterns, and then compose them together into - as you say - higher level abstractions, whose look and feel would support use cases - for instance, your sidecar example.

The gist is, I hope this PR is a step in that direction :) 

More to come in the near future.

Reply:
Yes, that is really good. And this comment is just stating my view. As I see both ways have their cons and pros.